components/cloud/src/blob.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use futures_io::AsyncRead;
+pub use kvproto::backup::CloudDynamic;
+use std::io;
+use std::marker::Unpin;
+
+pub trait BlobConfig: 'static + Send + Sync {
+    fn name(&self) -> &'static str;
+    fn url(&self) -> io::Result<url::Url>;
+}
+
+/// An abstraction for blob storage.
+/// Currently the same as ExternalStorage
+pub trait BlobStorage: 'static + Send + Sync {
+    fn config(&self) -> Box<dyn BlobConfig>;
+
+    /// Write all contents of the read to the given path.
+    fn put(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()>;
+
+    /// Read all contents of the given path.
+    fn get(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_>;
+}
+
+impl BlobConfig for dyn BlobStorage {
+    fn name(&self) -> &'static str {
+        self.config().name()
+    }
+
+    fn url(&self) -> io::Result<url::Url> {
+        self.config().url()
+    }
+}
+
+impl BlobStorage for Box<dyn BlobStorage> {
+    fn config(&self) -> Box<dyn BlobConfig> {
+        (**self).config()
+    }
+
+    fn put(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()> {
+        (**self).put(name, reader, content_length)
+    }
+
+    fn get(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_> {
+        (**self).get(name)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct StringNonEmpty(String);
+impl StringNonEmpty {
+    pub fn opt(s: String) -> Option<Self> {
+        if s.is_empty() { None } else { Some(Self(s)) }
+    }
+
+    pub fn opt2(s1: String, s2: String) -> Option<Self> {
+        Self::opt(s1).or_else(|| Self::opt(s2))
+    }
+
+    pub fn required_field(s: String, field: &str) -> io::Result<Self> {
+        Self::required_msg(s, &format!("field {}", field))
+    }
+
+    pub fn required_field2(s1: String, s2: String, field: &str) -> io::Result<Self> {
+        match Self::opt2(s1, s2) {
+            Some(sne) => Ok(sne),
+            None => Err(Self::error_required(&format!("field {}", field))),
+        }
+    }
+
+    pub fn required_msg(s: String, msg: &str) -> io::Result<Self> {
+        if !s.is_empty() {
+            Ok(Self(s))
+        } else {
+            Err(Self::error_required(&format!("Empty {}", msg)))
+        }
+    }
+
+    pub fn error_required(msg: &str) -> io::Error {
+        io::Error::new(io::ErrorKind::InvalidInput, msg)
+    }
+
+    pub fn required(s: String) -> io::Result<Self> {
+        Self::required_msg(s, "string")
+    }
+
+    pub fn stat(s: &'static str) -> Self {
```

Comment:
i'd rather you call it `static_` or `static_str` or `r#static`.

this name sounds like `stat(2)`.